add linear g-formula estimator

akelleh opened the pull request:
Hey @amit-sharma !

I've added this but have a few questions. I'll add them inline.

Best,
A

msftclas commented:
[![CLA assistant check](https://cla.opensource.microsoft.com/pull/badge/not_signed)](https://cla.opensource.microsoft.com/Microsoft/dowhy?pullRequest=29) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.opensource.microsoft.com/Microsoft/dowhy?pullRequest=29) before we can accept your contribution.<br/><br/>:x: akelleh [sign now](https://cla.opensource.microsoft.com/Microsoft/dowhy?pullRequest=29)<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.opensource.microsoft.com/check/Microsoft/dowhy?pullRequest=29) it.</sub>

amit-sharma commented on dowhy/do_why.py:
```diff
@@ -158,6 +158,49 @@ def estimate_effect(self, identified_estimand, method_name=None,
             )
         return estimate
 
+    def estimate_outcome(self, x, identified_estimand, method_name=None,
```
Like the idea of a "do" function that works like the do operator. But looks like it is not implemented yet. Maybe you can submit the g-formula file separately from these edits, which may come in a future commit?

amit-sharma commented on dowhy/causal_estimators/gformula_estimator.py:
```diff
@@ -0,0 +1,63 @@
+import logging
+
+
+class GFormulaEstimator(CausalEstimator):
```
currently, this looks very similar to the LinearRegressionEstimator class. consider merging with it and adding the  do method there?

akelleh commented:
Hey Amit! 

Refactored to incorporate your feedback. I've added the do operation to the causal model, raise NotImplementedError in the case that an estimand doesn't support it, and made a quick little demo to show how it works:
https://gist.github.com/akelleh/9b486a2208102476a1a7be770afe6a31

Note this is using the trick with the Robins g-formula outlined here under "how it works":
https://medium.com/@akelleh/causal-inference-with-pandas-dataframes-fc3e64fce5d

amit-sharma commented on dowhy/causal_estimators/linear_regression_estimator.py:
```diff
@@ -24,24 +24,42 @@ def __init__(self, *args, **kwargs):
         self.logger.info("INFO: Using Linear Regression Estimator")
         self.symbolic_estimator = self.construct_symbolic_estimator(self._target_estimand)
         self.logger.info(self.symbolic_estimator)
+        self._linear_model = None
```
yes, storing the model on the estimator makes sense. 
In your gist example though, it seems that the user also has to specify the estimator ("backdoor.linear_regression"). Would it be cleaner to have them estimate the effect first, and then use "do" operator. The idea is that the do will use the last used estimator by default, unless you specify it.

amit-sharma commented:
Thanks for clarifying. 

>  Some models have parameters that are estimators for the causal effect, like linear models, but not all do.
This is a good point, and also makes me think that the do method, as implemented, is really an estimation operation. And so, ideally, any estimation algorithm can be written in terms of calls to do. (e.g. E(y|do(x=1) - E(y|do(x=0).  This will need some work to completely integrate, so let's leave "estimate" and "do" as separate for now. 

So the natural order then is model->identify->do->estimate, where estimate is really using the same method that was used in do.

One note on the do operation: I'm thinking that we could make the default behavior to output the E[Y|do(x)], and then have an parameter switch to output a probability distribution instead. How does that sound?

Re:Side-note: Did not fully understand. Given Y and X, the identification step outputs the identified estimand. So we assume that there is only causal quantity of interest.

akelleh commented:
@amit-sharma  Great! I think the conversation will come up again when I try to add more do operations. Pearl talks about an abstract quantity, Q, being identifiable in a causal model. P(y|do(x)) and E[y|do(x)] certainly qualify, but so does the ATE, ATT, ATC, LATE, etc.

I wonder if the right abstraction might be model -> identify(Q) -> estimate, where Q can be a do op, a causal effect, or any other quantity that's identifiable with the chosen method.
